<a href="https://colab.research.google.com/github/rafaelacirino/exercicios/blob/main/Introducao_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Introduçao Ao Pandas**

## Exercícios de treino

<center><img src="https://raw.githubusercontent.com/rafaelacirino/exercicios/main/panda.jpeg?token=ALZALWVQFTAIDZOUPYZUZRC756LUU"height="300px"></center>

Pandas é uma biblioteca do Python cuja função é manipulação e análise de dados. Pandas pega um dado (em formato csv, xml etc) e cria um objeto em Python com linhas e coluna para que possa tratar e analisar esses dados. Cada coluna é chamada Series e o conjunto de Series é chamado DataFrame. 
 Ao importar o conjunto de dados usando o Pandas, podem ser realizadas as seguintes ações, por exemplo:

* Extrair informações estatísticas:
  * Qual a média, mediana, valores máximos e mínimos?
  * Qual é a distribuição das suas variáveis?
  * Qual a correlação entre duas variáveis quaisquer?
* Exportar os dados para um novo formato de arquivo
* Visualizar gráficos dos mais diferentes tipos
* Alimentar modelos de *machine learning* feitos em cima do Scikit-learn

# Instalar o Pandas


In [1]:
import pandas as pd

# Importar arquivo csv

Para importar o arquivo csv, utilizarei como exemplo os dados da ação ordinária COGN3.SA. Os dados foram extraídos do site [Yahoo Finance](https://finance.yahoo.com/), foi feito o download e disponibilizado o arquivo no seguinte [link](https://github.com/rafaelacirino/exercicios/blob/main/COGN3.SA.csv).
Para importar o arquivo é simples, basta digitar o seguinte comando:
df = pd.read_csv("link do arquivo").
O arquivo csv deve ser armazenado em uma variável que, neste caso, será chamada de df.

In [2]:
#importar arquivo csv
df = pd.read_csv("https://raw.githubusercontent.com/rafaelacirino/exercicios/main/COGN3.SA.csv?token=ALZALWRHBLX6H37ZBX36ADS756M2S")

In [3]:
#ver o tipo da variável 
type(df)

pandas.core.frame.DataFrame

In [4]:
#mostrar o tamanho do DataFrame
df.shape

(125, 7)

Aqui esta função nos mostrou qual o tamanho do DataFrame. Neste caso, ele é formado por 125 linhas e 7 colunas.

In [7]:
#mostrar o nome das colunas
df.columns

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')

Aqui mostra o nome das colunas e seu tipo. Neste caso, é do tipo objeto. Para saber o tipo de cada coluna, basta a função seguinte: type(df['nome_da_coluna']) ou df.dtypes que retornará o tipo de cada coluna

In [8]:
#saber o tipo da coluna
type(df['Close'])

pandas.core.series.Series

In [13]:
#saber o tipo de cada coluna
df.dtypes

Date          object
Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume         int64
dtype: object

Para saber se há valores ausentes e retornar o tipo de cada coluna, basta df.info()

In [14]:
#tipo e se há valores ausentes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125 entries, 0 to 124
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       125 non-null    object 
 1   Open       125 non-null    float64
 2   High       125 non-null    float64
 3   Low        125 non-null    float64
 4   Close      125 non-null    float64
 5   Adj Close  125 non-null    float64
 6   Volume     125 non-null    int64  
dtypes: float64(5), int64(1), object(1)
memory usage: 7.0+ KB


In [5]:
#mostrar as 5 primeiras linhas
df.head()

Date  Open  High   Low  Close  Adj Close     Volume
0  2020-07-01  6.64  6.97  6.59   6.97       6.97   56882300
1  2020-07-02  7.11  7.42  6.78   6.92       6.92   86721200
2  2020-07-03  6.93  7.20  6.80   7.20       7.20   34928400
3  2020-07-06  8.05  8.05  7.43   7.60       7.60  130989600
4  2020-07-07  7.53  7.78  7.32   7.45       7.45   77476000

In [6]:
#mostrar as 5 últimas linhas
df.tail()

Date  Open  High   Low  Close  Adj Close    Volume
120  2020-12-22  4.66  4.69  4.49   4.50       4.50  48485700
121  2020-12-23  4.52  4.63  4.52   4.60       4.60  31338200
122  2020-12-28  4.62  4.65  4.55   4.60       4.60  45061400
123  2020-12-29  4.63  4.84  4.62   4.69       4.69  52931100
124  2020-12-30  4.71  4.76  4.63   4.63       4.63  27328600

Fazer esses passos, df.head() e df.tail(), mostrará se o DataFrame seguiu um parâmetro, ou seja, como nosso DataFrame foi estruturado.

Vamos selecionar uma coluna específica e calcular algumas informações estatísticas

In [15]:
#selecionar e calcular média, mediana e somatório da coluna High
df['High'].mean()

6.21128

In [16]:
df['High'].median()

5.61

In [17]:
df['High'].sum()

776.4100000000001

#Trabalhando com Datas
Para trabalhar com datas deve-se primeiro converter o formato da coluna respectiva. Primeiro deve-se analisar o tipo (df.Date, mostra o tipo da coluna Date), geralmente o formato padrão é o object. Para converter e poder manipular os dados das datas, usa a seguinte função: df.Date = pd.to_datetime(df.Date, format = "%Y-%m-%d").dentro dos parêntesis, informa qual o nome da coluna (df.Date), e o formato que se encontra a data (neste caso está no formato americano - %Y-%m-%d - %Y pois o ano está em 4 dígitos, caso fossem 2, seria %y, separado por traços e depois %m, mês, %d, dia).  

In [18]:
#mostra o formato da coluna Date
df.Date

0      2020-07-01
1      2020-07-02
2      2020-07-03
3      2020-07-06
4      2020-07-07
          ...    
120    2020-12-22
121    2020-12-23
122    2020-12-28
123    2020-12-29
124    2020-12-30
Name: Date, Length: 125, dtype: object

In [19]:
#converter a coluna Date que está no tipo object para o tipo datetime
df.Date = pd.to_datetime(df.Date, format = '%Y-%m-%d')

In [20]:
#mostrar que foi modificado
df.Date

0     2020-07-01
1     2020-07-02
2     2020-07-03
3     2020-07-06
4     2020-07-07
         ...    
120   2020-12-22
121   2020-12-23
122   2020-12-28
123   2020-12-29
124   2020-12-30
Name: Date, Length: 125, dtype: datetime64[ns]

In [21]:
#acessa qualquer informação relacionado a data
df.Date.dt.year

0      2020
1      2020
2      2020
3      2020
4      2020
       ... 
120    2020
121    2020
122    2020
123    2020
124    2020
Name: Date, Length: 125, dtype: int64

In [22]:
#acessa mês
df.Date.dt.month

0       7
1       7
2       7
3       7
4       7
       ..
120    12
121    12
122    12
123    12
124    12
Name: Date, Length: 125, dtype: int64

In [23]:
#acessa a semana do ano
df.Date.dt.weekofyear

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  


0      27
1      27
2      27
3      28
4      28
       ..
120    52
121    52
122    53
123    53
124    53
Name: Date, Length: 125, dtype: int64